test_conformance/non_uniform_work_group/TestNonUniformWorkGroup.cpp

```diff
@@ -703,12 +703,46 @@ size_t TestNonUniformWorkGroup::adjustGlobalBufferSize(size_t globalBufferSize)
   err = clGetDeviceInfo(_device, CL_DEVICE_MAX_MEM_ALLOC_SIZE, sizeof(deviceMaxAllocObjSize), &deviceMaxAllocObjSize, NULL);
   test_error(err, "clGetDeviceInfo failed");
 
-  size_t adjustedGlobalBufferSize = globalBufferSize;
-  if (deviceMaxAllocObjSize < globalBufferSize) {
-    adjustedGlobalBufferSize = deviceMaxAllocObjSize;
-    log_info("globalBufferSize was adjusted from %lu to %lu\n", globalBufferSize, adjustedGlobalBufferSize);
+  if (deviceMaxAllocObjSize >= globalBufferSize) return globalBufferSize;
+
+  // Adjust _GlobalSize and _numOfGlobalWorkItems as well in order to avoid
+  // out-of-bound access of global buffer in testBarriers.
+  size_t maxNumOfGlobalWorkItems = deviceMaxAllocObjSize / sizeof(cl_uint);
+  size_t oldNumOfGlobalWorkItems = _numOfGlobalWorkItems;
+  size_t oldGlobalSize[MAX_DIMS];
+  memcpy(oldGlobalSize, _globalSize, sizeof(_globalSize));
+  for (int i = (int)_dims - 1; i >= 0; --i)
+  {
+      size_t minGlobalSize = _localSize_IsNull ? 1 : _localSize[i];
+      while (_numOfGlobalWorkItems > maxNumOfGlobalWorkItems
+             && _globalSize[i] >= (2 * minGlobalSize))
+      {
+          _globalSize[i] /= 2;
+          _numOfGlobalWorkItems =
+              _globalSize[0] * _globalSize[1] * _globalSize[2];
+      }
+  }
+  // If _globalSize[0] is adjusted, make sure there is still non-uniform
+  // workgroup.
+  if (_globalSize[0] != oldGlobalSize[0])
+  {
+      if (_globalSize[0] % _localSize[0] == 0)
+      {
+          _globalSize[0]--;
+          _numOfGlobalWorkItems =
+              _globalSize[0] * _globalSize[1] * _globalSize[2];
+      }
   }
 
+  size_t adjustedGlobalBufferSize = _numOfGlobalWorkItems * sizeof(cl_uint);
+  log_info("globalSize was adjusted from [%zu, %zu, %zu] to [%zu, %zu, %zu]\n",
+           oldGlobalSize[0], oldGlobalSize[1], oldGlobalSize[2], _globalSize[0],
+           _globalSize[1], _globalSize[2]);
+  log_info("numOfGlobalWorkItems was adjusted from %zu to %zu\n",
+           oldNumOfGlobalWorkItems, _numOfGlobalWorkItems);
+  log_info("globalBufferSize was adjusted from %zu to %zu\n", globalBufferSize,
+           adjustedGlobalBufferSize);
+
   return adjustedGlobalBufferSize;
```

Comment:
I have test it with really strange value `deviceMaxAllocObjSize == 164096` and subcase `non_uniform_other_basic`. I noticed that before `return` we have to run `calculateExpectedValues();`. 
Additionally there could be an issue with `_globalWorkOffset` which is not updated after modification `_globalSize`

Reply:
thank you. I've added calculateExpectedValues and adjusted _globalWorkOffset.